### PR TITLE
Pull request #114 addition

### PR DESCRIPTION
**Pull request by goto40**
Playing around for TEP-001 I refactored the solution for #103 to be more generic.

**Changes**
Instead of hard code the importURI/_tx_loaded_models attriute when resolving a FQN, I 
allowed to provide a scope_redirection_logic to replace a current object in an FQN chain.

**Effect**
With this, #103 is implemented. However it also allows to define package-aliases (see scoping.md and the regression test for #103).

**Docu**
This work also helped to **document, why the Postponed mechanism is not part of the FQN** (docu  adapted). 

Explanation text (no open point):
"It remains unclear if a processed unresolved element (Postponed) will be relevant for the current name. This **would** easily create circular dependencies for the current FQN implementation. Therefore, Postponed is excluded (rational included in scoping.md)."

### CODE REVIEW

File: tests/functional/regressions/test_issue103_python_like_import.py [hunk 1]
```diff
@@ -50,3 +52,105 @@ def importURI_to_scope_name(import_obj):
...
+            package p2 {
+                class a {};
+            }
+        }
+        using p1.p2 as main
+        var x = new p1.p2.a()
+        var y = new main.a()
+    ''')
+
+    # first test (negativ example)
```
Comment (igordejanovic) [inline tests/functional/regressions/test_issue103_python_like_import.py]: typo: `negativ` -> `negative`

File: tests/functional/regressions/test_issue103_python_like_import.py [hunk 2]
```diff
@@ -50,3 +52,105 @@ def importURI_to_scope_name(import_obj):
...
+                package p2 {
+                    class a {};
+                }
+            }
+            using p1.p2 as main
+            var x = new p1.p2.a()
+            var y = new error.a()
+        ''')
+
+    # first test (negativ example)
```
Comment (igordejanovic) [inline tests/functional/regressions/test_issue103_python_like_import.py]: typo: `negativ` -> `negative`

File: textx/scoping/providers.py [hunk 3]
```diff
@@ -153,6 +162,18 @@ def _find_obj_fqn(p, fqn_name, cls):
             """
 
             def find_obj(parent, name):
+                if parent is not current_obj and \
```
Comment (igordejanovic) [inline textx/scoping/providers.py:166]: Do we need `current_obj` parameter in all these function. As I see it carries the value of top level `obj` and is never changed. Could we just say here `if parent is not obj and ...`  and remove all those `current_obj` params?

File: textx/scoping/providers.py [hunk 4]
```diff
@@ -201,20 +218,20 @@ def _find_referenced_obj(p, name, cls):
...
+            ret = _find_obj_fqn(p, name, cls, current_obj)
             if ret:
                 return ret
             while hasattr(p, "parent"):
                 p = p.parent
-                ret = _find_obj_fqn(p, name, cls)
+                ret = _find_obj_fqn(p, name, cls, current_obj)
                 if ret:
                     return ret
             return None
```
Comment (igordejanovic) [inline textx/scoping/providers.py]: I've just noticed this. In Python if function don't return anything it is implied that it returns `None`.

File: textx/scoping/providers.py [hunk 5]
```diff
@@ -201,20 +218,20 @@ def _find_referenced_obj(p, name, cls):
...
 
         from textx.model import ObjCrossRef
         assert type(obj_ref) is ObjCrossRef, type(obj_ref)
         obj_cls, obj_name = obj_ref.cls, obj_ref.obj_name
-        ret = _find_referenced_obj(obj, obj_name, obj_cls)
+        ret = _find_referenced_obj(obj, obj_name, obj_cls, obj)
         if ret:
             return ret
         else:
```
Comment (igordejanovic) [inline textx/scoping/providers.py]: Can we just say here `return _find_referenced_obj(obj, obj_name, obj_cls)` ? 
No need for `if` afterwards I would say.

**Review by igordejanovic — CHANGES_REQUESTED** [review]
Looks good. Sorry for taking me so long. 
I have some minor changes to discuss. Please see inline comments.

**Comment by goto40** [comment]
Thank you for the feedback. It is ok for to leave some pull requests aside until we have time to discuss (this is no central feature)... I will work on the comments now.

File: tests/functional/regressions/test_issue103_python_like_import.py (hunk 1 above)
Comment (goto40) [reply tests/functional/regressions/test_issue103_python_like_import.py]: fixed

File: tests/functional/regressions/test_issue103_python_like_import.py (hunk 2 above)
Comment (goto40) [reply tests/functional/regressions/test_issue103_python_like_import.py]: fixed

File: textx/scoping/providers.py (hunk 3 above)
Comment (goto40) [reply textx/scoping/providers.py:166]: Absolutely! I removed the superfluous parameters. However I changed the top-level parameter name to "current_obj" to make the code more readable.

File: textx/scoping/providers.py (hunk 4 above)
Comment (goto40) [reply textx/scoping/providers.py]: OK. Nice to know. I removed two "return None" but also added one comment at the end of the while loop indicating that an implicit "return None" is expected in some cases.

File: textx/scoping/providers.py (hunk 5 above)
Comment (goto40) [reply textx/scoping/providers.py]: Ok. See last comment (this is the second return None which I removed).

**Comment by goto40** [comment]
Changed code according to the comments.

**Comment by goto40** [comment]
Note: I noticed that we can "Squash and merge" instead of "Merge pull request" in the github interface (you need to drop down the green button). This might make it easier for other commiters to just say "merge and squash" my stuff into  the master...

**Comment by igordejanovic** [comment]
Thanks. Looks good. Merging.

> Note: I noticed that we can "Squash and merge" instead of "Merge pull request" in the github interface (you need to drop down the green button). This might make it easier for other commiters to just say "merge and squash" my stuff into the master...

Yes, that can be used if all commits should be squashed but if finer control is needed then git interactive rebase is the way to go.

**Pull request merged**